packages/opentelemetry-api/src/common/Exception.ts

```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Defines Exception.
+ *
+ * Error or string
+ */
+export type Exception = Partial<Error> | string;
```

Comment:
because this is what spec says we allow for an error object which has at least either type or message otherwise we should not record

Reply:
Well from my understanding [the spec says](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#record-exception) that `The signature of the method is to be determined by each language and can be overloaded as appropriate.` with the following *examples*:

- RecordException(exception: Exception)
- RecordException(type: String, message: String, stacktrace: String)

Considering that its up to us, i would be in favor of accepting either a `string` or an `Error`

Reply:
@vmarchaud here -> https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/exceptions.md#recording-an-exception

Reply:
Not sure i get what are you trying to show, you linked the attributes of the event not the spec concerning the signature of the `recordException` method

Reply:
yes we have to record error type or message - this is the minimum requirement. So we can't require whole Error - as the message or type is already enough for us to be able to log something. that's why the type of exception is `Partial<Error> | message`

Reply:
I see your point, however in my understanding it's not required to allow to only give the message or only gives the type.

Reply:
but why ?, this is exactly what is written in spec `One of exception.type or exception.message is required`

Reply:
You can still follow that line in the spec even without `Partial`. That is specifying the minimum we must export, but has nothing to do with what is passed into our helper function. With a `Partial<Error> | string`, it would be possible to pass something to our function which we cannot export. With `Error | string`, we know we can always export.

Reply:
```
interface MyOwnError1 {
  type: string
}
interface MyOwnError2 {
  message: string
}

const error1: MyOwnError1 = { type = 'foo' };
const error2: MyOwnError2 = { message = 'foo' };

// this should be possible, but after removing `Partial` it won't be as you get an error
span.recordException(error1);
span.recordException(error2);

// and the same without defining your own interface
span.recordException({type: 'foo'});
span.recordException({type: 'foo', message: 'foo message'});
```

these all different cases should be possible without any error

Reply:
I argue that they should not (and thats not required by the spec)

Reply:
Why would it be possible to call `recordException` with something that is not an exception or a string? The way I read the spec, we don't have to make any such compromises. We just have to make sure that we are adding `type` or `message` at least to the event attributes. I also don't see a strong use-case for that.

Reply:
I think I would like to remove `Partial`, and do something like this:

```typescript
recordException(exception: Error | string, startTime?: TimeInput) {
  if (typeof exception === 'string') {
    return span.addEvent('exception', { 'exception.message': exception }, startTime);
  }

  // Unfortunately, we cannot simply check instanceof Error https://github.com/microsoft/TypeScript/issues/13965
  if (
    typeof exception.code === 'string' &&
    typeof exception.message === 'string' &&
    typeof exception.stack === 'string'
  ) {
    return span.addEvent('exception', {
      'exception.type': exception.code,
      'exception.message': exception.message,
      'exception.stacktrace': exception.stack,
    }, startTime);
  }

  // user may not be using typescript and API may be used incorrectly
  logger.warn(`Span.recordException called with invalid value ${exception}`);
}
```

Implementation is simpler and we are following the spec (exception events always have at least `code` or `message`.

Reply:
`exception events always have at least code or message.`

There are many different kind of errors / exceptions either in node or in web. Some of them might have more params or less. type "Error" is very generic and spec says:

`The method MUST record an exception as an Event with the conventions outlined in the exception semantic conventions document.`
Here are semantic "exception semantic conventions"
https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/exceptions.md#attributes

Where we can read about 3 params
1. exception.type
`Required?` One of exception.type or exception.message is required
2. exception.message
`Required?` One of exception.type or exception.message is required
3. exception.stacktrace
`Required?` No

So our "Error" interface has 3 params `type` or `message` where one of them is required, and `stacktrace` is optional - not required

so things like
```
span.recordError({type: 'foo'});
span.recordError({message: 'foo'});
span.recordError({message: 'foo', stacktrace:'dasdasdas'});
span.recordError({type: 'aaaa', stacktrace:'dasdasdas'});
span.recordError({message: 'foo', type: 'aaaa', stacktrace:'dasdasdas'});
```
are all valid cases

And as last what is worth to mention that you will not always record certain error directly 
```
try {
} catch(e) {
  span.recordException(e);
}

```
You might quite often need to map fields that are required from the error that happened
```
// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypeError
try {
} catch(e: TypeError) {
  span.recordException({type: e.name, message: e.message});
}
```
which is perfectly valid case.

Reply:
The most accurate Exception type should look like this
```
interface ExceptionWithType {
  message?: string;
  stacktrace?: string;
  type: string;
}
interface ExceptionWithMessage {
  message: string;
  stacktrace?: string;
  type?: string;
}
/**
 * Defines Exception.
 *
 * string or an object with one of (message or type) and optional stacktrace
 */
export type Exception = ExceptionWithType | ExceptionWithMessage | string;
```

Reply:
So you're proposing to put the burden to translate whatever error object the plugin sees to this format before calling recordException, then what is the purpose of having  this helper function anyways if it doesn't actually do anything different than the regular `addEvent`?

Reply:
I was wrong with naming params for interfaces - was just writing from out of my head - updated

Reply:
so if certain error has required field you can still do this
```
try{} catch(e){
span.recordException(e);
}
```

Reply:
here
https://github.com/open-telemetry/opentelemetry-js/pull/1372/commits/3ed3c40cb3acdd75a0bea8c651dd28dc0b3c9410

Reply:
> so things like
span.recordException({type: 'foo'});
span.recordException({message: 'foo'});
span.recordException({message: 'foo', stacktrace:'dasdasdas'});
span.recordException({type: 'aaaa', stacktrace:'dasdasdas'});
span.recordException({message: 'foo', type: 'aaaa', stacktrace:'dasdasdas'});
are all valid cases

I believe you are confusing what are the required attributes on the `exception` event versus what are required on the signature. Only the former is in the spec, we have the choice on the later.

> so if certain error has required fields you can still do this
try{} catch(e){
span.recordException(e);
}

This should not compile in Typescript because `e` is `any` or `unknown` (depending on the typescript version) so i don't think the example is valid.

@obecny could you please explain whats the benefits of allowing custom error objects ? Still don't see the value instead of just allowing `Error | string`,

Reply:
I agree with @vmarchaud here. I think `recordException(exception: Error | string, time?: api.TimeInput)` would be fine.

If exception is a string, set the `exception.message` event attribute. If the exception is an error, set the `exception.type`, `exception.message`, and `exception.stacktrace` event attributes.

Both of these cases are spec-compliant since they contain the minimum event attributes. This is also consistent with what other SIGs have

Reply:
I think we are going in circles. @dyladan your comment from 8 days ago https://github.com/open-telemetry/opentelemetry-js/pull/1372#discussion_r464708776 , it has been already refactored to include this case too. interface  "Exception" has nothing to do with interface "Error" - the interface which even lacks the mentioned `code` field. We can talk on chat as I don't want to repeat again the same the arguments from over a week which leads to the linked message and the conclusion to refactor I did then.

Reply:
>I think we are going in circles. @dyladan your comment from 8 days ago #1372 (comment) , it has been already refactored to include this case too.

>interface "Exception" has nothing to do with interface "Error" - the interface which even lacks the mentioned code field

The `e.code` field is unrelated to this, it was just something I was suggesting we map to the `exception.type` attribute. `e.name` is also fine if this is a problem for you.

The `Exception` interface does not need to exist. `Error` is missing the `code` field because the `@types/node` package is missing it. There is an issue open for this, but it hasn't gotten any traction: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/42623. `code` has existed since node 8.

I think you are confusing fields required to be input to the API and fields required to be added as event attributes. The specification specifically allows that the arguments to `recordException` may be whatever makes sense for the language in question. I believe that `Error | string` makes sense for this language. The required `exception.message`/`exception.type` are event attributes, not inputs for this method.

That is what @vmarchaud was saying when he said this:

>I believe you are confusing what are the required attributes on the exception event versus what are required on the signature. Only the former is in the spec, we have the choice on the later.